lib/spirv/lib/src/instructions.dart

```diff
@@ -1,34 +1,287 @@
 part of spirv;
 
+class _BlockContext {
+  final StringBuffer out;
+  final int indent;
+
+  /// The most local merge block id, or zero.
+  final int merge;
+
+  /// The most local continue block id, or zero.
+  final int continueBlock;
+
+  /// The most local loop-construct header block id.
+  final int loopHeader;
+
+  /// The most local loop-construct merge block id.
+  /// This is different from [merge] when the context is inside an if-statement
+  /// inside of a for-loop, for example.
+  final int loopMerge;
+
+  _BlockContext({
+    required this.out,
+    required this.indent,
+    this.merge = 0,
+    this.continueBlock = 0,
+    this.loopHeader = 0,
+    this.loopMerge = 0,
+  });
+
+  /// Return a new [_BlockContext] that is a copy of the current [_BlockContext]
+  /// with an increased indent and any parameters specified here overwritten.
+  _BlockContext child({
+    int? merge,
+    int? continueBlock,
+    int? loopHeader,
+    int? loopMerge,
+  }) =>
+      _BlockContext(
+        out: out,
+        indent: indent + 1,
+        merge: merge ?? this.merge,
+        continueBlock: continueBlock ?? this.continueBlock,
+        loopHeader: loopHeader ?? this.loopHeader,
+        loopMerge: loopMerge ?? this.loopMerge,
+      );
+
+  void writeIndent() {
+    out.write('  ' * indent);
+  }
+}
+
 class _Block {
   List<_Instruction> instructions = <_Instruction>[];
 
   void add(_Instruction i) {
     instructions.add(i);
   }
 
-  void writeIndent(StringBuffer out, int indent) {
-    for (int i = 0; i < indent; i++) {
-      out.write('  ');
+  final int id;
+  final _Function function;
+
+  _Block(this.id, this.function);
```

Comment:
See the style guide for this and some other classes on how to order members: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#constructors-come-first-in-a-class